aidb/engine/approx_select_engine.py

```diff
@@ -0,0 +1,225 @@
+
+
+  def get_sampled_proxy_blob(self, proxy_score_for_all_blobs, defensive_mixing: int = 0.1):
+    weights = proxy_score_for_all_blobs.values ** 0.5
+    normalized_weights = (1 - defensive_mixing) * (weights / sum(weights)) + defensive_mixing / len(weights)
+    mass = 1 / len(weights) / normalized_weights
+
+    dataset = pd.DataFrame({
+        PROXY_SCORE: proxy_score_for_all_blobs.values,
+        WEIGHT: normalized_weights,
+        MASS: mass},
+        index=proxy_score_for_all_blobs.index
+    )
+
+    return dataset
+
+
+  def tau_estimate(self, recall_target, satisfied_sampled_results):
+    estimated_tau = 0.0
+    recall_score = 0
+
+    for index, blob in satisfied_sampled_results.iterrows():
+      recall_score += blob[MASS]
+      if recall_score >= recall_target:
+        estimated_tau = blob[PROXY_SCORE]
+        break
+    return estimated_tau
+
+
+  def _get_confidence_bounds(self, mu, sigma, s, delta):
+    if s == 0:
+      return 0.0, 0.0
+    val = (sigma / math.sqrt(s)) * math.sqrt(2 * math.log(1 / delta))
+    return mu - val, mu + val
+
+
+  def tau_modified(self, satisfied_sampled_results, recall_target, confidence, total_length):
+    z = satisfied_sampled_results[MASS]
+    estimated_tau = self.tau_estimate(
+      recall_target * sum(z),
+      satisfied_sampled_results
+    )
+    positive_length = len(satisfied_sampled_results[satisfied_sampled_results[PROXY_SCORE] >= estimated_tau])
+
+    estimated_z1 = list(z[:positive_length]) + [0] * (total_length - len(z[:positive_length]))
+    estimated_z2 = list(z[positive_length:]) + [0] * (total_length - len(z[positive_length:]))
+    z1_mean, z1_std = np.mean(estimated_z1), np.std(estimated_z1)
+    z2_mean, z2_std = np.mean(estimated_z2), np.std(estimated_z2)
+
+    delta = 1.0 - confidence
+
+    _, ub = self._get_confidence_bounds(z1_mean, z1_std, len(estimated_z1), delta / 2)
+    lb, _ = self._get_confidence_bounds(z2_mean, z2_std, len(estimated_z2), delta / 2)
+
+    # inflate recall to correct the sampling errors
+    if (ub+lb) == 0.0:
+      modified_recall_target = 1.0
+    else:
+      modified_recall_target = ub / (ub + lb)
+
+    modified_tau = self.tau_estimate(
+      modified_recall_target * sum(z),
+      satisfied_sampled_results
+    )
+
+    logger.info(f'modified_tau: {modified_tau}')
+    return modified_tau
+
+
+  async def execute_approx_select_query(self, query: Query):
+    if not query.is_valid_approx_select_query:
+      raise Exception('Approx select query should contain Confidence and should not contain Budget.')
+
+    # generate proxy score for each blob
+    score_for_all_df, score_connected = await self.get_proxy_scores_for_all_blobs(query)
+
+    proxy_score_for_all_blobs = self.score_fn(score_for_all_df, score_connected)
+
+    dataset = self.get_sampled_proxy_blob(proxy_score_for_all_blobs)
+
+    # This is used for parallel test
+    seed = (mp.current_process().pid * np.random.randint(100000, size=1)[0]) % (2**32 - 1)
```

Comment:
Can you pass in the seed as an argument to the base engine? I don't think it's good to set here